### PR TITLE
WIP: Add Doozer verb to build images with Konflux

### PR DESCRIPTION
**Pull request by vfreex**

### CODE REVIEW

File: doozer/doozerlib/backend/konflux_image_builder.py [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+    def _new_component(name: str, application: str, component_name: str,
+                       image_repo: Optional[str], source_url: Optional[str], revision: Optional[str]) -> dict:
+        obj = {
+            "apiVersion": "appstudio.redhat.com/v1alpha1",
+            "kind": "Component",
+            "metadata": {
+                "name": name,
+                "annotations": {
+                    "build.appstudio.openshift.io/pipeline": '{"name":"docker-build","bundle":"latest"}',
+                }
```
Comment (ashwindasr) [inline doozer/doozerlib/backend/konflux_image_builder.py:184]:
```
mintmaker.appstudio.redhat.com/disabled=true
```
Need to add this annotation in, just to be safe. ref [thread](https://redhat-internal.slack.com/archives/GDBRP5YJH/p1724260588819889?thread_ts=1723123622.362349&cid=GDBRP5YJH)

**Review by ashwindasr — CHANGES_REQUESTED** [review]

File: doozer/doozerlib/backend/konflux_image_builder.py [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+
+
+class KonfluxImageBuildError(Exception):
+    def __init__(self, message: str, pipelinerun_name: str, pipelinerun: Optional[resource.ResourceInstance]) -> None:
+        super().__init__(message)
+        self.pipelinerun_name = pipelinerun_name
+        self.pipelinerun = pipelinerun
+
+
+KONFLUX_BUNDLE_PULLSPEC = "quay.io/konflux-ci/tekton-catalog/pipeline-docker-build:f3ac40bbc0230eccb8d98a4d54dabd55a4943c5d"
```
Comment (ashwindasr) [inline doozer/doozerlib/backend/konflux_image_builder.py:29]: Looks like that SHA is pretty outdated. We probably need to figure out a process, but for now we can probably use the latest build always. But I see that there is `devel` so can we try something like

`quay.io/konflux-ci/tekton-catalog/pipeline-docker-build:devel`

File: doozer/doozerlib/backend/konflux_image_builder.py [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+        app_name = self._config.group_name.replace(".", "-")
+        app_manifest = self._new_application(app_name, app_name)
+        app = await self._create_or_patch(dyn_client, app_manifest)
+        self._logger.info(f"Created application: {app['metadata']['name']}")
+
+        # Ensure the component resource exists
+        component_name = f"{app_name}-{metadata.distgit_key}"
+        dest_image_repo = self._config.output_repo
+        # dest_image_tag = f"{metadata.distgit_key}_{self._uuid_tag}"
+        dest_image_tag = f"{version}-{release}"
```
Comment (thegreyd) [inline doozer/doozerlib/backend/konflux_image_builder.py:110]: why not name-version-release?

File: doozer/doozerlib/backend/konflux_image_builder.py [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+            "source_url": https_url,
+            "revision": commit_sha,
+            "target_branch": target_branch,
+            "git_auth_secret": git_auth_secret,
+
+        })
+        obj = yaml.load(rendered)
+        del obj["metadata"]["name"]
+        del obj["metadata"]["namespace"]
+        del obj["metadata"]["annotations"]["pipelinesascode.tekton.dev/on-cel-expression"]
```
Comment (thegreyd) [inline doozer/doozerlib/backend/konflux_image_builder.py:234]: curious what's the reason behind this / or why do we have this in yaml in the first place if we don't want them?

File: doozer/doozerlib/backend/konflux_image_builder.py [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+            await exectools.to_thread(api.get, name=name, namespace=namespace)
+        except exceptions.NotFoundError:
+            found = False
+        if self._config.dry_run:
+            if found:
+                self._logger.warning(f"[DRY RUN] Would have patched {api_version}/{kind} {namespace}/{name}")
+            else:
+                self._logger.warning(f"[DRY RUN] Would have created {api_version}/{kind} {namespace}/{name}")
+            return resource.ResourceInstance(dyn_client, manifest)
+        if found:
```
Comment (thegreyd) [inline doozer/doozerlib/backend/konflux_image_builder.py:267]:
```suggestion
        if found:
            self._logger.info(f"Patching {api_version}/{kind} {namespace}/{name}")
```

File: doozer/doozerlib/backend/konflux_image_builder.py [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+            new = await exectools.to_thread(
+                api.patch,
+                body=manifest,
+                namespace=namespace,
+                content_type="application/merge-patch+json",
+            )
+        else:
```
Comment (thegreyd) [inline doozer/doozerlib/backend/konflux_image_builder.py:274]:
```suggestion
        else:
            self._logger.info(f"Creating {api_version}/{kind} {namespace}/{name}")
```

File: doozer/doozerlib/backend/rebaser.py [hunk 7]
```diff
@@ -0,0 +1,1549 @@
...
+        if self._runtime.assembly:
+            sb.write(".assembly.")
+            sb.write(self._runtime.assembly)
+
+        el_ver = 0
+        try:
+            el_ver = metadata.branch_el_target()
+        except ValueError:
+            pass
+        if el_ver:
```
Comment (thegreyd) [inline doozer/doozerlib/backend/rebaser.py:545]: We've grown to depend on `.elX` suffix in our nvr names, so i think we should error out if it is not found instead of passing, so we can investigate

File: doozer/doozerlib/backend/rebaser.py [hunk 8]
```diff
@@ -0,0 +1,1549 @@
...
+        self._add_build_repos(dfp)
+
+        self._reflow_labels(df_path)
+
+    def _add_build_repos(self, dfp: DockerfileParser):
+        # Populating the repo file needs to happen after every FROM before the original Dockerfile can invoke yum/dnf.
+        dfp.add_lines(
+            "\n# Start Konflux-specific steps",
+            "RUN mkdir -p /tmp/yum_temp; mv /etc/yum.repos.d/*.repo /tmp/yum_temp/ || true",
+            "COPY .oit/signed.repo /etc/yum.repos.d/",
```
Comment (thegreyd) [inline doozer/doozerlib/backend/rebaser.py:718]: I think we want to change this to be signed/unsigned depending on the version 

https://github.com/openshift-eng/art-tools/blob/24fd4652d8ef4b803b28c21c3d0d1f9aa6d21340/doozer/doozerlib/osbs2_builder.py#L161

File: doozer/doozerlib/backend/rebaser.py (hunk 7 above)
Comment (thegreyd) [reply doozer/doozerlib/backend/rebaser.py:545]: but can also do in a separate PR, so upto you if you wanna make this change here

File: doozer/doozerlib/constants.py [hunk 9]
```diff
@@ -29,6 +29,12 @@
 KONFLUX_REPO_CA_BUNDLE_TMP_PATH = "/tmp"
 KONFLUX_REPO_CA_BUNDLE_FILENAME = "Current-IT-Root-CAs.pem"
 KONFLUX_REPO_CA_BUNDLE_HOST = "https://certs.corp.redhat.com/certs"
+WORKING_SUBDIR_KONFLUX_BUILD_SOURCES = "konflux_build_sources"
```
Comment (ashwindasr) [inline doozer/doozerlib/constants.py:32]: Noticed that these constants aren't being used? Or not sure if I missed something or it'll be in an another PR.

File: doozer/doozerlib/backend/konflux_image_builder.py (hunk 3 above)
Comment (vfreex) [reply doozer/doozerlib/backend/konflux_image_builder.py:110]: You are right. Initially by default config each image gets an separate repo. Since we are going to use a mono repo for all images, we need to include the image name.

File: doozer/doozerlib/backend/konflux_image_builder.py (hunk 4 above)
Comment (vfreex) [reply doozer/doozerlib/backend/konflux_image_builder.py:234]: In the future the PipelineRun template file (https://github.com/openshift-eng/art-tools/pull/859/files#diff-f1894aed6cfd6dbe45cd9701265dbf035239886d026b4dca0c40865f5e8511fe) will be loaded from an external repo, which includes some fields that are specific to a single image. We'd better to remove them.

Comment (vfreex) [reply doozer/doozerlib/backend/konflux_image_builder.py:234]: I'll add a comment here

File: doozer/doozerlib/backend/rebaser.py (hunk 7 above)
Comment (vfreex) [reply doozer/doozerlib/backend/rebaser.py:545]: Right. This part was taken from existing code and I didn't change it. Let's make a change in future PRs.

File: doozer/doozerlib/backend/rebaser.py (hunk 8 above)
Comment (vfreex) [reply doozer/doozerlib/backend/rebaser.py:718]: This part was taken from original Ashwin's PR. Let's address this in a separate PR.

File: doozer/doozerlib/constants.py (hunk 9 above)
Comment (vfreex) [reply doozer/doozerlib/constants.py:32]: right, it is used in https://github.com/openshift-eng/art-tools/pull/820

**Comment by vfreex** [comment]
@thegreyd @ashwindasr Thanks for the review. I've addressed your review suggestions. Please have another look.
Note previously this PR includes some commits from the rebase PR https://github.com/openshift-eng/art-tools/pull/820, so it was a bit confusing. For those review suggestions, I will open another PR to address them.
